File: JAVA/longestSubstring.java
import java.util.HashSet;
// Unordered collections of unique elements that does not allow unique characters
// Efficeient insertation, deletion and lookup


class Solution {
    public int lengthOfLongestSubstring(String s) {
        // Declaration for the lengthOfLongestSubstring method
        // Returns int but takes in a single string


        int n = s.length(); // Calculates length of input string to n
        int i = 0; // left pointer of the window
        int j = 0; // right pointer of window
        int maxLength = 0;


        HashSet<Character> set = new HashSet<>();
        // HashSet<Character> set: set variable will store objects of type Character in a HashSet
        // OR it creates a HashSet that can store characters
        // new HashSet<>(): creates a new hashset free of any elements


        while (j < n) { // continue until all loops are finished
            if (!set.contains(s.charAt(j))) {
                // if is to check if current character at index is present already
                // !set.contains(s.charAt(j)): if it is not present we add it to set using set.add(s.charAt(j));
                set.add(s.charAt(j));
                maxLength = Math.max(maxLength, j - i + 1);
                // Update maxLength
                j++;
                // Increment j by one to expand the window

            } else { // if character is already present in set we have encountered a repeatinc character, sooo
                set.remove(s.charAt(i));
                i++;
                // set.remove(s.charAt(i));: remove the character at index i from set
                // i++: increment left side of window 
            }
        }

        return maxLength;
    }
}

File: JavaScript/longestSubstring.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let maxLength = 0;
    let left = 0;
    let right = 0;
    let set = new Set();
    // new Set(): is a built in object that only stores unique values- duplicate values will be eliminated
    // Set maintains the order of insertion
    // Set also has a few methods: add(), delete(), has() --> check existence & a size property

    while(right < s.length) {
        if (!set.has(s.charAt(right))) {
            // "if the character at the current position in the string is not already present in the set.."
            // .has method checks if value is present 
            // charAt() method is used to access the character at a specific index
            // So it checks if the character at the current position is unique, which allows us to add it to the set

            set.add(s.charAt(right));
            // "Add the character at the current positon in the string to the set"
            // Adds the unique characters to the current substring

            maxLength = Math.max(maxLength, set.size);
            right++;
            // Update the max length and move the window over a slot
        } else {

            // If the character at the current postion is present ...
            // Then delete the character at the left point and move the window
            set.delete(s.charAt(left));
            left++;
        }
    }

    return maxLength;
};
